### PR TITLE
AccountDelete.DeletedBy and PackageDelete.DeletedBy should be set to null when the User is deleted

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/NuGet/Engineering/issues/2358

### CODE REVIEW

File: src/NuGet.Services.Entities/AccountDelete.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ public class AccountDelete
         /// <summary>
         /// The User(admin) key that executed the delete action.
         /// </summary>
-        public int DeletedByKey { get; set; }
+        public int? DeletedByKey { get; set; }
```
Comment (loic-sharma) [inline src/NuGet.Services.Entities/AccountDelete.cs]: >DeletedByKey [](start = 20, length = 12)

Please add a comment as to why this row may be null. Same thing for `PackageDelete.DeletedByKey`

Comment (shishirx34) [inline src/NuGet.Services.Entities/AccountDelete.cs]: will be null for self deletes?

File: src/NuGetGallery/Services/DeleteAccountService.cs [hunk 2]
```diff
@@ -109,6 +112,7 @@ private async Task DeleteAccountImplAsync(User userToBeDeleted, User userToExecu
             await RemoveSecurityPolicies(userToBeDeleted);
             await RemoveUserCredentials(userToBeDeleted);
             await RemovePackageOwnershipRequests(userToBeDeleted);
+            ResetPackagesAndAccountsDeletedBy(userToBeDeleted);
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/DeleteAccountService.cs]: >ResetPackagesAndAccountsDeletedBy [](start = 12, length = 33)

What would happen if the account deletes a package after `ResetPackagesAndAccountsDeletedBy` is called but before `SaveChangesAsync`?

File: src/NuGet.Services.Entities/AccountDelete.cs (hunk 1 above)
Comment (scottbommarito) [reply src/NuGet.Services.Entities/AccountDelete.cs]: Right now this is set to the site admin who executes the user delete. When this flow is automated, it depends on how it is implemented by @ryuyu

Comment (scottbommarito) [reply src/NuGet.Services.Entities/AccountDelete.cs]: Done.

File: src/NuGetGallery/Services/DeleteAccountService.cs (hunk 2 above)
Comment (scottbommarito) [reply src/NuGetGallery/Services/DeleteAccountService.cs]: The deletion process happens in a single transaction so it's not possible for a `User` delete a package (or do any action) at any point in the deletion process.

File: src/NuGetGallery/Services/DeleteAccountService.cs [hunk 3]
```diff
@@ -318,6 +322,25 @@ private void RemoveMembers(Organization organization)
...
         }
 
+        private void ResetPackagesAndAccountsDeletedBy(User user)
+        {
+            foreach (var deletedPackage in _packageDeleteRepository
+                .GetAll()
+                .Where(d => d.DeletedByKey == user.Key)
+                .ToList())
+            {
+                deletedPackage.DeletedBy = null;
```
Comment (zhhyu) [inline src/NuGetGallery/Services/DeleteAccountService.cs]: Don't you also clean up "DeleteByKey"?

File: tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs [hunk 4]
```diff
@@ -141,6 +141,8 @@ public async Task DeleteHappyUser(bool isPackageOrphaned, AccountDeletionOrphanP
                     Assert.NotEmpty(testUser.OrganizationMigrationRequests);
                     Assert.NotEmpty(testUser.OrganizationRequests);
                     Assert.NotEmpty(testUser.Organizations);
+                    Assert.NotNull(testableService.PackageDeletedByUser.DeletedBy);
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs]: I think you should also check that "DeletedByKey" is set to null or not for all these test cases?

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs [hunk 5]
```diff
@@ -356,7 +356,7 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
                 .HasRequired(a => a.DeletedAccount);
 
             modelBuilder.Entity<AccountDelete>()
-                .HasRequired(a => a.DeletedBy)
+                .HasOptional(a => a.DeletedBy)
```
Comment (zhhyu) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: Do we have any test cases to cover this change?

**Comment by zhhyu** [comment]
Should we have any functional tests to cover the whole picture of deleting accounts?

File: tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs [hunk 6]
```diff
@@ -533,6 +544,8 @@ public DeleteAccountTestService(User user, PackageRegistration userPackagesRegis
                     NewOwner = _user
                 });
 
+                AccountDeletedByUser = new AccountDelete { DeletedBy = _user, DeletedByKey = _user.Key };
```
Comment (zhhyu) [inline tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs]: Maybe it's better that we include some test cases which contain empty "AccountDeletedByUser" or "PackageDeletedByUser" to improve test coverage? It's just a suggestion, not a must.

**Comment by zhhyu** [comment]
Should we have any functional tests/E2E tests to cover the whole picture of deleting accounts including this change?

File: src/NuGetGallery/Services/DeleteAccountService.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/Services/DeleteAccountService.cs]: If you set one, EF will set the other automatically when you commit changes.

**Comment by scottbommarito** [comment]
@zhhyu - we have wanted to add functional tests for this, but unfortunately it is difficult to make a test around this primarily because deleting accounts requires being a site admin, and there are difficulties associated with giving a test account site admin.

File: tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs (hunk 4 above)
Comment (scottbommarito) [reply tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs]: I'd rather not, for two reasons:
- If one is set to null, the other must be set to the same thing or EF is not working properly
- `DeletedByKey` won't even be set to null because the EF behavior that sets it to null does not execute in unit tests

File: tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs (hunk 6 above)
Comment (scottbommarito) [reply tests/NuGetGallery.Facts/Services/DeleteAccountServiceFacts.cs]: Good idea. Will add!
